Add the Pyright type checker

akselaase opened the pull request:
This PR introduces tooling for the Pyright Python type checker, usable both as a local tool during development, and also integrated into the CI pipeline.

The CI check works in a similar way to the current `flake8-annotations` check, in that it compares the number of typing errors before and after applying an incoming pull request. However, the existing check only counts the number of missing annotations, which does not help in ensuring correctness of the annotations. For instance, [this parameter](https://github.com/rucio/rucio/blob/58efd21b5e21182df80bef3dbe8befa636e440b8/lib/rucio/core/transfer.py#L581) should probably not be a `List[str]`, but a `str`, which Pyright rightly notices in the function body.

The CI check prevents new PRs from increasing the error count. The check also attempts to intelligently recognize where the new errors were introduced, to help the developer find their mistakes more easily. However, this could cause a disruption in other developers' workflow, so maybe some guidelines for how strictly we adhere to the CI result should be developed. Pyright's config can also be modified to be more relaxed if it proves to be too noisy. 

Pyright was chosen due to the relative ease of configuring it and processing its output compared to MyPy. Other issues like MyPy limiting the amount of output in case of too many errors also led to the PR focusing on Pyright (though this issue could maybe be resolved by passing the `--soft-error-limit N` flag to MyPy).

Another benefit of Pyright (that could just as well be present in other type checkers - I didn't look) is the ability to integrate it into your IDE of choice through the Language Server Protocol. I know at least VSCode has seamless integration when using the [Pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance) extension, which greatly improves the developer experience by highlighting errors immediately, without having to wait for the CI check to run.

As a follow-up to this PR, I want to gradually improve type annotations and correctness throughout the codebase. Pyright reports numerous issues, some of which could be bugs that need fixing.

Closes #5727

joeldierkes commented on tools/run_pyright/__main__.py:
```diff
@@ -0,0 +1,54 @@
+    parser = ArgumentParser(description="Runs the pyright type checker.")
+    subparser = parser.add_subparsers(dest='command')
+    subparser.required = True
+    func: Callable[[Namespace], int]
```
I don't see the value of this line, `func` is always initialized.

akselaase replied:
It's to type it properly. Ensures that errors show up if any of the `{generate,compare,...}.setup_parser()` calls doesn't return the correct type.

joeldierkes commented on tools/run_pyright/compare.py:
```diff
@@ -0,0 +1,93 @@
+def setup_parser(parser: ArgumentParser) -> Callable[[Namespace], int]:
+    parser.add_argument('old', type=Path, help='First report of comparison.')
+    parser.add_argument('new', type=Path, help='Second report of comparison.')
+    parser.add_argument('--werror', action='store_true', help='Treat warnings as errors.')
```
I only know this argument from gcc, there it is `--Werror`...

joeldierkes commented on tools/run_pyright/compare.py:
```diff
@@ -0,0 +1,93 @@
+    print_regressions(new_errors, new_report)
+
+    any_errors = any(err.severity == 'error' for err, _ in new_errors)
+    any_warnings = any(err.severity == 'warning' for err, _ in new_errors)
```
`serverity` could be an enum. Thoughts?

joeldierkes commented on .github/workflows/autotest.yml:
```diff
@@ -71,6 +71,41 @@ jobs:
+        shell: bash
+        run: |
+          tools/run_pyright.sh generate \
+            ${{ env.UPDATED_PYRIGHT_REPORT_FILE }}
```
I want to see the output of the GitHub Actions job. Should we treat warnings here as errors? (`--werror`)

akselaase replied:
Both reports are now shown in a separate step in the workflow.

joeldierkes commented on tools/run_pyright/generate.py:
```diff
@@ -0,0 +1,79 @@
+    if args.werror:
+        num_errors += report.summary.num_warnings
+
+    return 1 if num_errors > 0 else 0
```
While I like return codes, I am not completely sure if this one is right. The doc comment says `Generate a pyright-report and save it at the specified path.`, which implies for me that a non-zero return code results in a not created pyright-report, not that pyright found some errors. Thoughts?

joeldierkes commented on tools/run_pyright/generate.py:
```diff
@@ -0,0 +1,79 @@
+
+def setup_parser(parser: ArgumentParser) -> Callable[[Namespace], int]:
+    parser.add_argument('out', type=Path, help='Store Pyright report at this path.')
+    parser.add_argument('--strip', choices=['line', 'character', 'range'], help='Strip certain attributes from the output.')
```
An explanation of why we need that would be nice, not necessarily in the git history in that case ;-)

akselaase replied:
At the moment unused, but I initially used it to produce reports that could be diffed without getting too much noise from changing line numbers. Unsure whether to leave it there or not, thoughts? If so, will explain it better.

akselaase replied:
Agreed to remove since it is unused.

joeldierkes commented on tools/run_pyright/models.py:
```diff
@@ -0,0 +1,90 @@
+            range_start_char=obj['range']['start']['character'],
+            range_end_line=obj['range']['end']['line'],
+            range_end_char=obj['range']['end']['character'],
+        )   # type: ignore
```
Number of spaces before comment

joeldierkes commented on tools/run_pyright/utils.py:
```diff
@@ -0,0 +1,46 @@
+        return json.load(f)
+
+
+def save_json(out: Union[Path, IO[str]], data: Dict[str, Any]) -> None:
```
`save_json` is called once with a `Path`. Is the recursive call and the split of the parameter type better compared to:
```python
def save_json(path: Path, data: Dict[str, Any]) -> None:
    with open(path, 'w') as f:
        json.dump(data, f, indent=4)
        f.write('\n')
```

joeldierkes commented on .github/workflows/autotest.yml:
```diff
@@ -71,6 +71,41 @@ jobs:
             exit 1
           fi
+  python_pyright:
+    name: Check Pyright
```
Maybe a more descriptive name than the package, like `Check Types`/`Check Static Typing`?

akselaase replied:
How about "Python type check (Pyright)"?

joeldierkes commented on tools/run_pyright/generate.py:
```diff
@@ -0,0 +1,79 @@
+def _run_pyright() -> ReportDict:
+    """Runs the pyright type-checker and returns its output as json."""
+    cmdline = ['pyright', '--outputjson', *PATHS]
+    process = subprocess.run(cmdline, stdout=subprocess.PIPE)
```
The error message in case `pyright` is not installed is not descriptive:
```
...
  File "/usr/lib/python3.10/subprocess.py", line 1845, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'pyright'
```
I'd suggest to either give more information on how to resolve that, or maybe use `npx`?

joeldierkes reviewed (changes requested):
While I like `mypy`, I don't mind using some other tool. In this case some more information would be nice:
- Why don't we use mypy?
- Why pyright? There are other tools like facebooks [`pyre-ceck`](https://github.com/facebook/pyre-check) and googles [`pytype`](https://github.com/google/pytype).

Very good PR, this helps a lot! Small nits/Questions.

akselaase commented on .github/workflows/autotest.yml:
```diff
@@ -71,6 +71,45 @@ jobs:
+        run: |
+          tools/run_pyright.sh compare \
+            ${{ env.PYRIGHT_ANCESTOR_REPORT }} \
+            ${{ env.PYRIGHT_CURRENT_REPORT }}
```
To decide: With or without `--Werror`?

akselaase replied:
Added `--Werror`. Better to get the reviewer to decide in case something is questionable.

akselaase commented on .github/actions/checkout_ancestor_commit/action.yaml:
```diff
@@ -53,6 +59,12 @@ runs:
+          mkdir -p ../tmp/$(dirname $f)
+          cp -r $f ../tmp/$f
+        done
+
```
Could also just rename `files_to_copy` to `paths_to_copy` and have it support both files and directories by just adding `-r` to the `cp` command, instead of having separate loops for files and directories.

joeldierkes replied:
That's the cleaner alternative, go for it.
Please create a proper commit with a proper commit message for that (including the intention).

You can force push the branch and amend/edit your changes. The comments on GitHub are still present in the view (if the line is not deleted) and you have revisions for each review.

akselaase commented on tools/run_pyright/__main__.py:
```diff
@@ -0,0 +1,54 @@
+sys.stdout = codecs.getwriter('utf-8')(sys.stdout.buffer)
+sys.stderr = codecs.getwriter('utf-8')(sys.stderr.buffer)
+
+
```
Not needed when were setting `PYTHONIOENCODING=utf8` I believe, todo: test.

akselaase commented on tools/run_pyright/generate.py:
```diff
@@ -0,0 +1,95 @@
+        print('Unknown error running pyright.\n'
+              'Additional details:', ex,
+              file=sys.stderr)
+        raise
```
This doesn't really provide anything useful.

akselaase committed:
CI: Support copying directories in action "checkout_ancestor_commit"

Previously only files could be copied due to the `cp` command being
called without `--recursive`. This commit adds the recursive option,
and renames the workflow input from `files_to_copy` to `paths_to_copy`
to better reflect its support for copying directories.

akselaase committed:
Tools: Pyright static type checking; Fix #5727

This commit introduces the Pyright type checker to the CI pipeline. The
CI check works in a similar way to the current "flake8-annotations"
check, in that it compares the number of typing errors before and after
applying an incoming pull request. The check fails if the number of
errors has increased. The added value of this check is to check for
typing errors, instead of just counting missing annotations as in the
"flake8-annotations" check. These two checks are orthogonal to each
other and serve different purposes. Pyright to a certain degree does
type inference, but produces better results in presence of type
annotations. The annotations themselves serve little value if there is
no type checker to use them except for increased readability. Therefore,
the "flake8-annotations" check will continue to be active.

akselaase commented:
Demo of output can be seen in the "Compare reports" step here: https://github.com/rucio/rucio/runs/7469292269?check_suite_focus=true

joeldierkes commented on tools/run_pyright/compare.py:
```diff
@@ -0,0 +1,105 @@
+            else:
+                yield prefix + line
+            first = False
+    return ''.join(_inner())
```
nit: maybe we can utilize the `textwrap.indent` function for that?
Something like:
```python
lines = text.splitlines(keepends=True)
return lines[0] + textwrap.indent(''.join(lines[1:]), prefix, lambda_: True)
```

joeldierkes reviewed (dismissed):
Good PR!

nit: The second commit message is not wrapped.
nit: The phrase `Pyright would have a hard time without any annotations` in the commit message is rather generic. Maybe it should be explained here that Pyright is more of a "static" type checker and requires some type annotations. While other type checkers are more dynamic and better at type inference, they did not provide the needed functionalities and integration into tooling. (or are just painfully slow).

joeldierkes commented on tools/run_pyright/compare.py:
```diff
@@ -0,0 +1,97 @@
+
+def _indent(text: str, prefix: str) -> str:
+    """Prepends `prefix` to each line in `text` except the first."""
+    return text.replace('\n', '\n' + prefix)
```
:joy:

akselaase commented:
Slightly edited the commit message and simplified the indentation function. This dismissed the review, is this desired behaviour? It can be configured as shown here: https://github.com/isaacs/github/issues/1157#issuecomment-450409941.

joeldierkes commented:
I don't know if I'n in favor of that. This use case hardly ever appears, and to resolve it you just have to re-request a new review.

Since we work on an open-source project, I'd rather have explicit approval than someone who changes more than the requested changes.

akselaase commented:
To make sure this works in the development container, https://github.com/rucio/containers/pull/197 should also be merged. I can prepare a separate PR for documentation.

bari12 commented:
Needs to be rebased due to an issue which existed on master yesterday.

joeldierkes commented:
I think this is good to go :+1:
